src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -91,27 +95,51 @@ class BlockFetcher(
 
   private def handleHeadersMessages(state: BlockFetcherState): Receive = {
     case Response(peer, BlockHeaders(headers)) if state.isFetchingHeaders =>
-      log.debug("Fetched {} headers starting from block {}", headers.size, headers.headOption.map(_.number))
-
-      val newState = state.validatedHeaders(headers) match {
-        case Left(err) =>
-          peersClient ! BlacklistPeer(peer.id, err)
-          state.fetchingHeaders(false)
-        case Right(validHeaders) =>
-          state.appendHeaders(validHeaders)
-      }
+      val newState =
+        if (state.fetchingHeadersState == AwaitingHeadersToBeIgnored) {
+          log.debug(
+            "Received {} headers starting from block {} that will be ignored",
+            headers.size,
+            headers.headOption.map(_.number)
+          )
+          state.withHeaderFetchReceived
+        } else {
+          log.debug("Fetched {} headers starting from block {}", headers.size, headers.headOption.map(_.number))
+
+          state.validatedHeaders(headers) match {
+            case Left(err) =>
+              peersClient ! BlacklistPeer(peer.id, err)
+              state.withHeaderFetchReceived
+            case Right(validHeaders) =>
+              state.withHeaderFetchReceived.appendHeaders(validHeaders)
+          }
+        }
 
       fetchBlocks(newState)
     case RetryHeadersRequest if state.isFetchingHeaders =>
-      log.debug("Retrying request for headers")
-      fetchHeaders(state)
+      log.debug("Time-out occurred while waiting for headers")
```

Comment:
more of discussion type comment. Do you think it would bring problems if we would receive this time out response later ? i.e something like:
1. We request header lets say from block with number 100
2. we receive invalidate from 100 to 80
3. we receive timeout for our request, we blacklist peer and clear state.
4. after some time (200 s by default) peer is unblacklisted , and we request headers from block with number 90 form it.
5.  we receive old response due to some network traffic problems or something like that.

Do you think it could result in something more than blacklising peer ? (for example triggering this bug with unbounded resources usage)

Reply:
Definitely, we are currently doing no validations that the headers received match the request

With that sort of validations we should be safe I think, but for now I was assuming that that case will never happen, the 30 seconds till timeout should be more than enough to prevent that, right?

Reply:
That makes me think if this whole solution shouldn't be implemented as Actor/class in between fetcher and peers client. It could track requests made and invalidate previous ones basing on strategy choosen (at-most-n, allow-all, etc.). 
This could simplify things significantly and allow code in fetcher focus on just orchestrating fetching of blocks.

Reply:
What would this actor do? Are you maybe thinking of something like the BlockChainDataFetcher of our other project? (only on how it requests headers and bodies)

>It could track requests

That at least would focus on :point_up:  part and allow validating that the responses received match the requests we sent

Either way I'd do that effort in parallel with this kind of patching that only touches the minimum required (and is needed for the testnet)

I'm a bit worried of how risky any design changes on our sync could be... though that BlockChainDataFetcher refactoring would be low risk I think

Reply:
Well, `BlockFetcher` is more or less the same as `BlockchainDataFetcher` there already IMO. I think more on structure like that:
```
                           block fetcher
                      |       |                      |
headers fetcher(Fetcher)  bodies fetcher(Fetcher)  state nodes fetcher(Fetcher)
                      |       |                      |
                             peers client
```
Where `Fetcher` is that specialized class/actor that manages requests of given type and could have interface like:
```
abstract class Fetcher(multipleReqStrategy: AtMostN|AllowAll) {
  def makeRequest(msg: MessageSerializable): IO[msg.Response] //Issues a request and add it to tracked pool
  def cancelPendingRequests: IO[Unit] //Stores information about cancellation so knows which responses cannot be passed further if received
}  
```

Reply:
so imo details up to discussion, but i am all for doing some kinda detailed request-response tracking. Unfortunately as eth do not have any request-id for its request-response messages, such thing is ultimately necessary if we want to have possiblitly to cancel in fly requests.